### PR TITLE
basic TNS submission

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR adds basic TNS submission support.

### CODE REVIEW

File: static/js/components/TNSATForm.jsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  tnsrobotSelectItem: {
+    whiteSpace: "break-spaces",
+  },
+  container: {
+    width: "99%",
+    marginBottom: "1rem",
+  },
+}));
+
+const TNSATForm = ({ obj_id }) => {
```
Comment (guynir42) [inline static/js/components/TNSATForm.jsx:34]: What does the AT stand for in TNSATForm?

File: static/js/components/TNSATForm.jsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  // render to update it, so it can be null even if allocationList is not
+  // empty.
+  if (tnsrobotList.length === 0 || !selectedTNSRobotId) {
+    return <h3>No TNS robots available...</h3>;
+  }
+
+  const handleSubmit = async ({ formData }) => {
+    setSubmissionRequestInProcess(true);
+    formData.tnsrobotID = selectedTNSRobotId;
+    console.log("obj_id", obj_id);
```
Comment (guynir42) [inline static/js/components/TNSATForm.jsx:82]:
```suggestion
```

File: static/js/components/TNSSpectraForm.jsx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    // Don't want to reset everytime the component rerenders and
+    // the defaultStartDate is updated, so ignore ESLint here
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dispatch, setSelectedTNSRobotId]);
+
+  // need to check both of these conditions as selectedAllocationId is
+  // initialized to be null and useEffect is not called on the first
+  // render to update it, so it can be null even if allocationList is not
+  // empty.
```
Comment (guynir42) [inline static/js/components/TNSSpectraForm.jsx:74]: I think you need to update this comment about allocations.

File: static/js/ducks/source.js [hunk 4]
```diff
@@ -14,6 +14,8 @@ const ADD_CLASSIFICATION = "skyportal/ADD_CLASSIFICATION";
 
 const DELETE_CLASSIFICATION = "skyportal/DELETE_CLASSIFICATION";
 
+const ADD_TNS = "skyportal/ADD_TNS";
```
Comment (guynir42) [inline static/js/ducks/source.js:17]: Maybe better to use `ADD_SOURCE_TNS` and `addSourceTNS`?

File: static/js/ducks/spectra.js [hunk 5]
```diff
@@ -17,6 +17,8 @@ const PARSE_SOURCE_SPECTRUM_ASCII_OK =
 
 export const RESET_PARSED_SPECTRUM = "skyportal/RESET_PARSED_SPECTRUM";
 
+const ADD_TNS = "skyportal/ADD_TNS";
```
Comment (guynir42) [inline static/js/ducks/spectra.js:20]: Same here with spectrum. In fact, I'm not sure if this is a bug where multiple API calls have the same tag. In any case, better make them specific.

File: skyportal/handlers/api/tns.py [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+              type: string
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+        """
+        tnsrobot = TNSRobot.get_if_accessible_by(
+            int(tnsrobot_id), self.current_user, mode='delete'
+        )
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:157]:
```suggestion
        try:
            tnsrobot_id = int(tnsrobot_id)
        except ValueError:
            return self.error("TNSRobot ID must be an integer.")
        tnsrobot = TNSRobot.get_if_accessible_by(
            tnsrobot_id, self.current_user, mode='delete'
        )
```

File: skyportal/handlers/api/tns.py [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+        altdata = tnsrobot.altdata
+        if not altdata:
+            raise ValueError('Missing TNS information.')
+
+        tns_headers = {
+            'User-Agent': 'tns_marker{"tns_id":'
+            + str(tnsrobot.bot_id)
+            + ', "type":"bot", "name":"'
+            + tnsrobot.bot_name
+            + '"}'
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:220]: This can be an f-string.

File: skyportal/handlers/api/tns.py [hunk 8]
```diff
@@ -0,0 +1,592 @@
...
+        }
+
+        time_first, mag_first, magerr_first, filt_first, instrument_first = (
+            None,
+            None,
+            None,
+            None,
+            None,
+        )
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:229]:
```suggestion
        time_first = mag_first = magerr_first = filt_first = instrument_first = None
```
Although there's a chance one of the lint/black formatters will not like this...

File: skyportal/handlers/api/tns.py [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+        (
+            time_last_nondetection,
+            limmag_last_nondetection,
+            filt_last_nondetection,
+            instrument_last_nondetection,
+        ) = (None, None, None, None)
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:242]: Same here. But please first check that it passes formatting.

File: skyportal/handlers/api/tns.py [hunk 10]
```diff
@@ -0,0 +1,592 @@
...
+        }
+
+        at_report = {
+            "ra": {"value": obj.ra},
+            "dec": {"value": obj.dec},
+            "groupid": tnsrobot.source_group_id,
+            "internal_name_format": {
+                "prefix": "ZTF",
+                "year_format": "YY",
+                "postfix": "",
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:309]: This is ZTF specific.

File: skyportal/handlers/api/tns.py [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+            },
+            "internal_name": obj.id,
+            "reporter": reporters,
+            "discovery_datetime": astropy.time.Time(time_first, format='mjd').jd,
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:313]: Should this be `jd` or `datetime`?

File: skyportal/handlers/api/tns.py [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+
+        altdata = tnsrobot.altdata
+        if not altdata:
+            raise ValueError('Missing TNS information.')
+
+        spectrum = Spectrum.get_if_accessible_by(
+            spectrum_id,
+            self.current_user,
+            raise_if_none=True,
+        )
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:423]: Catch if doesn't find anything.

File: skyportal/handlers/api/tns.py [hunk 13]
```diff
@@ -0,0 +1,592 @@
...
+        external_observer = (
+            DBSession()
+            .query(SpectrumObserver.external_observer)
+            .filter(SpectrumObserver.spectr_id == spectrum_id)
+            .first()
+        )
+        if external_observer is not None:
+            spec_dict["external_observer"] = external_observer[0]
+
+        self.verify_and_commit()
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:450]: Why is this here?

File: skyportal/handlers/api/tns.py [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+
+        tns_headers = {
+            'User-Agent': 'tns_marker{"tns_id":'
+            + str(tnsrobot.bot_id)
+            + ', "type":"bot", "name":"'
+            + tnsrobot.bot_name
+            + '"}'
+        }
+
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:459]: f-string

File: skyportal/handlers/api/tns.py [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+            spectrum_type
+        ) + 1
+
+        if spec_dict["altdata"] is not None:
+            header = spec_dict["altdata"]
+            observation_date = (
+                str(header['OBSDATE'].split('T')[0])
+                + ' '
+                + str(header['OBSDATE'].split('T')[1])
+            )
```
Comment (guynir42) [inline skyportal/handlers/api/tns.py:479]: f-string

**Review by guynir42 — COMMENTED** [review]
Looks good. I have some small comments.

File: static/js/components/TNSATForm.jsx (hunk 1 above)
Comment (mcoughlin) [reply static/js/components/TNSATForm.jsx:34]: Astronomical transient

File: static/js/ducks/source.js (hunk 4 above)
Comment (mcoughlin) [reply static/js/ducks/source.js:17]: good idea.

File: static/js/ducks/spectra.js (hunk 5 above)
Comment (mcoughlin) [reply static/js/ducks/spectra.js:20]: yes good.

File: static/js/components/TNSATForm.jsx (hunk 1 above)
Comment (guynir42) [reply static/js/components/TNSATForm.jsx:34]: Should have known that...

File: skyportal/handlers/api/tns.py (hunk 13 above)
Comment (mcoughlin) [reply skyportal/handlers/api/tns.py:450]: nonsense

File: skyportal/handlers/api/tns.py (hunk 12 above)
Comment (mcoughlin) [reply skyportal/handlers/api/tns.py:423]: yup.

File: skyportal/handlers/api/tns.py (hunk 15 above)
Comment (mcoughlin) [reply skyportal/handlers/api/tns.py:479]: yup

File: skyportal/handlers/api/tns.py (hunk 14 above)
Comment (mcoughlin) [reply skyportal/handlers/api/tns.py:459]: yup.

File: skyportal/handlers/api/tns.py (hunk 10 above)
Comment (mcoughlin) [reply skyportal/handlers/api/tns.py:309]: oops

**Review by guynir42 — APPROVED** [review]
LGTM.

**Comment by mcoughlin** [comment]
Thanks @guynir42.
